### ISSUE TITLE
Can I send ciphertext by using binary IO?

### ISSUE THREAD

**Issue by Jyun-Neng** [homenc/HElib#229]
If I encrypt my data, then I want to send this ciphertext to server. 
Can I use binary IO to send ciphertext?
I know there is way to send secret key, public key, and context base by using binary IO.
But if I can send ciphertext by using binary IO?

**Comment by PhantomR** [comment]
Yes! I think what you want is the Ctxt::write and Ctxt::read functions, though you shoud look into the >> and << operators as well.. I'm not exactly sure which is binary and which is ASCII. 

I used write & read like this (using this wonderful example https://github.com/OpenSMP/SMP/blob/master/test/test_net_io.cpp):

On the Client, I had a Ctxt like:
`Ctx ct(publicKey);
publicKey.Encrypt(ct, to_ZZX(1));
ct.write(outputStream);`

On the Server, I created a new ciphertext:
Ctxt readCt(publicKey);
ct.read(inputStream);

If you get any kinds of errors trying this method, please also take into account this thread: https://github.com/shaih/HElib/issues/226 , where I actually had some apparent problems by using write & read compared to >> and <<, though, as I mentioned there, I am NOT sure it wasn't due to me forgetting to rebuild code!

**Comment by PhantomR** [comment]
@Kualdra can you provide the code for both the client and server sides? What did you use to compare two ciphertexts? Maybe the equality operator? I do not know how that is implemented

**Comment by Kualdra** [comment]
If you send something and get it back as the same, they should be equal, ciphered or not. For those after calculation yeah you are write may be! anyway unencrypted data should be always equal, even after calculation. 
That's for the client : 
`
    vector<Ctxt> ct1; // vecteur de ciphertexte
    for (int i = 0; i < 32; i++)
    {
        Ctxt cti(publicKey);
        ct1.push_back(cti);
    }
    publicKey.Encrypt(ct1[0], to_ZZX(v1[0]));
     connect << ct1[0];
     connect.flush();
     //can add some computation on ct1[0], same as on server
     //This resCompare vector is build like ct1
      connect >> resCompare1[0];
     //then compare encrypted and unencrypted data
     if(resCompare1[0] == ct1[0])
     {
          cout<<"they are equal encrypted"<<endl;
     }
     ZZX element1, element2;
     secretKey.Decrypt(element1, resCompare1[0]);
     secretKey.Decrypt(element2, ct1[0]);
     if( element1 == element2 )
     {
         cout<<"they are the same unencrypted"<<endl;
     }
`

And for the server side : 
`
            vector<Ctxt> Salaire;
            for (int i = 0; i < 32; i++) // On récupere le salaire du client
            {
                Ctxt cti(publicKey);
                Salaire.push_back(cti);
            }
             client >> Salaire[0];
             //can add some computation here, same as for client
             client << Salaire[0];
            client.close();
`

**Comment by shaih** [comment]
Comparing ciphertexts as in `resCompare1[0] == ct1[0]` is mostly a debugging operation, you should not expect it to provide semantically meaningful result.

**Comment by Kualdra** [comment]
Yes actually it is only for dubugging, i used it because i saw nonsense result when my server was doing computation for my client. But i still don't find out why :(

**Comment by shaih** [comment]
The most likely cause for nonsensical results is decryption errors. So the 1st thing to try is to use way more levels in context.buildModChain than you currently do and see if this fixes the problem.

**Comment by Kualdra** [comment]
i've tried many value for x in 
`buildModChain(context, L, c, x);`
but none of them give me a working decryption. But it looks like the problem is in the transmission more than the decryption step.

**Comment by fionser** [comment]
try larger `L`

**Comment by Kualdra** [comment]
i can't really increase L since i have limited RAM for my calculation after :(
